File: src/leetCode/java/Solution216.java
package leetCode.java;

import java.util.*;

//Backtracking O(k * (9! / (k! * (9 - k)!))) O(n), C9取k = 9! / (k! * (9 - k)!)
class Solution216 {
    public List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum3(int k, int n) {
        backtracking(1, new ArrayList<>(), k, n);
        return res;
    }

    public void backtracking(int start, List<Integer> nums, int k, int n){
        if(nums.size() == k && n == 0){
            res.add(new ArrayList<>(nums));
            return;
        }
        if(nums.size() > k || n < 0) return;

        for(int i=start; i<=9; i++){
            nums.add(i);
            backtracking(i + 1, nums, k, n - i);
            nums.remove(nums.size() - 1);
        }
    }
}

File: c++/Solution216.cpp
//Backtracking O(k * (9! / (k! * (9 - k)!))) O(n), C9取k = 9! / (k! * (9 - k)!)
class Solution216 {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> selected;
        backtracking(1, selected, k, n);
        return res;
    }

    void backtracking(int start, vector<int>& selected, int k, int target) {
        if (target == 0 && selected.size() == k) {
            res.push_back(selected);
            return;
        }
        if (target < 0) return;
        if (selected.size() >= k) return;

        for (int i = start; i <= 9; ++i) {
            selected.push_back(i);
            backtracking(i + 1, selected, k, target - i);
            selected.pop_back();
        }
    }
};